src/PhotonBackground.cpp

```diff
@@ -123,4 +125,293 @@ std::string photonFieldName(PhotonField photonField) {
+		} else {
+			ss2 = 26.4 * std::pow(s, -0.34) + 59.3 * std::pow(s, 0.095);
+		}
+		cs_multidiff = (1. - std::exp(-ss1)) * ss2;
+		cs_multi = 0.89 * cs_multidiff;
+		// diffractive scattering:
+		cross_diffr1 = 0.099 * cs_multidiff;
+		cross_diffr2 = 0.011 * cs_multidiff;
+		cross_diffr = 0.11 * cs_multidiff;
+		// **************************************
+		ss1 = std::pow(x - 0.85, 0.75) / 0.64;
+		ss2 = 74.1 * std::pow(x, -0.44) + 62. * std::pow(s, 0.08);
+		double cs_tmp = 0.96 * (1. - std::exp(-ss1)) * ss2;
+		cross_diffr1 = 0.14 * cs_tmp;
+		cross_diffr2 = 0.013 * cs_tmp;
+		double cs_delta = cross_frag2
+						- (cross_diffr1+cross_diffr2-cross_diffr);
+		if (cs_delta < 0.) {
+			cross_frag2 = 0.;
+			cs_multi += cs_delta;
+		} else {
+			cross_frag2 = cs_delta;
+		}
+		cross_diffr = cross_diffr1 + cross_diffr2;
+		cs_multidiff = cs_multi + cross_diffr;
+	// in the original SOPHIA code, here is a switch for the return argument.
+	// Here, only one case (compare in SOPHIA: NDIR=3) is needed.
+	}
+	return cross_res + cross_dir + cs_multidiff + cross_frag2;
+}
+
+double PhotonFieldSampling::Pl(double x, double xth, double xmax, double alpha) const {
+	if (xth > x)
+		return 0.;
+	const double a = alpha * xmax / xth;
+	const double prod1 = std::pow((x - xth) / (xmax - xth), a - alpha);
+	const double prod2 = std::pow(x / xmax, -a);
+	return prod1 * prod2;
+}
+
+double PhotonFieldSampling::Ef(double x, double th, double w) const {
+	const double wth = w + th;
+	if (x <= th) {
+		return 0.;
+	} else if ((x > th) && (x < wth)) {
+		return (x - th) / w;
+	} else if (x >= wth) {
+		return 1.;
+	} else {
+		throw std::runtime_error("error in function Ef");
+	}
+}
+
+double PhotonFieldSampling::breitwigner(double sigma_0, double Gamma, double DMM, double epsPrime, bool onProton) const {
+	const double mass = onProton? 0.93827 : 0.93947;  // Gev/c^2
+	const double s = mass * mass + 2. * mass * epsPrime;
+	const double gam2s = Gamma * Gamma * s;
+	return sigma_0 * (s / epsPrime / epsPrime) * gam2s / ((s - DMM * DMM) * (s - DMM * DMM) + gam2s);
+}
+
+
+double PhotonFieldSampling::functs(double s, bool onProton) const {
+	const double mass = onProton? 0.93827 : 0.93947;  // Gev/c^2
+	double factor = s - mass * mass;
+	double epsPrime = factor / 2. / mass;
+	double sigma_pg = crossection(epsPrime, onProton);
+	return factor * sigma_pg;
+}
+
+double PhotonFieldSampling::gaussInt(std::string type, double A, double B, bool onProton, double E_in, double z_in) const {
+	static const double X[8] = {.0950125098, .2816035507, .4580167776, .6178762444,
+						 .7554044083, .8656312023, .9445750230, .9894009349};
+	static const double W[8] = {.1894506104, .1826034150, .1691565193, .1495959888,
+						 .1246289712, .0951585116, .0622535239, .0271524594};
+	const double XM = 0.5 * (B + A);
+	const double XR = 0.5 * (B - A);
+	double SS = 0.;
+	for (int i = 0; i < 8; ++i) {
+		double DX = XR * X[i];
+		if (type == "functs") {
```

Comment:
Sorry that I missed this before, but the type-switch should not be  a string comparison, and certainly not in the inner loop! If you consider the gauss legendre integration temporary code, either use function pointers or simply create two methods.

If the integral will stay there, add a comment + file an issue (Feature request, gauss legendre integration) that this should go to a separate utility class, potentially from a external library. 

If only temporary, also add a comment that this is a temporary solution. Clearly marking temporaries is probably also good for the future reviews.

Reply:
Here, and elsewhere in the code, we should use enum class since it increases readability of a code, and makes less chance for run-time errors. This is a good practice coming to C++ from programming languages with more advanced type systems.

For example:
in `PhotonBackground.h`:
```c++
enum class IntegrationTypeFunction { functs, prob_eps };
```
and the interface becomes:
```c++
gaussInt(IntegrationTypeFunction type, ...
```
so in `PhotonBackground.cpp`:
```c++
if (type == IntegrationTypeFunction::functs)
```
Finally, with this approach you don't need the second else as any other cases besides those defined in the enum class are impossible and caught during the compile time.

However, in this particular case, I would suggest splitting the code as Tobias wrote.

Reply:
I have implemented a suggestion of how the Gauss integration is moved to `.../Common.h` while accepting arbitrary functions over which to integrate. However, using explicit function pointers did not work for me as the functions over which to integrate do not have the same signature.
* at this point: is there a way to define a variable function type to pass to a function over which to integrate?
* My current attempt in a402b8eb2460793a0ed9f79266f50e7694b6b402 tries to solve this based on [this post](https://stackoverflow.com/questions/15959374/c-pass-function-with-arbitrary-number-of-parameters-as-a-parameter). However, the compiler throws an error where the Gauss integration is called mentioning "invalid use of non-static member functions". I could not fix this but maybe there is a trivial solution to this that I am not seeing? If yes then we would have resolved all concerns evolving around the Gauss integration.

Reply:
I would do it like standard integration methods in other languages are doing it (see for example, python's [scipy](https://docs.scipy.org/doc/scipy/reference/generated/scipy.integrate.quad.html)): wrap your functions within another function (like a delta function), and provide it to a simplified gaussInt() interface which accepts only: f(x), A, B:

```c++
#include <functional>
(...)
double gaussInt(std::function<double(double)> integrand, double A, double B);
```

The next step is to rewrite `prob_eps` and `functs` so that they are [referential transparent](https://en.wikipedia.org/wiki/Referential_transparency), but I think, after a brief lookup, that they already are. Then, you should be able to call the integrator with a function something like (I haven't compiled it):

```c++
gaussInt(std::bind(
     [this, onProton, E_in, z_in](double x) { return this->prob_eps(x, onProton, E_in, z_in); }),
     epsMin, epsMax)
(...)
gaussInt(std::bind([this, onProton](double x){ return this->functs(x, onProton); }), smin, smax)
```

I'll check exact pieces tomorrow.

Reply:
Sorry, calling goes without std::bind():
```c++
gaussInt(
     [this, onProton, E_in, z_in](double x) { return this->prob_eps(x, onProton, E_in, z_in); },
     epsMin, epsMax);
(...)
gaussInt(
     [this, onProton](double x){ return this->functs(x, onProton); }),
     smin, smax);

Move gaussInt() to `Common.cpp`:
```c+
double gaussInt(std::function<double(double)> integrand, double A, double B) {
        static const double X[8] = {.0950125098, .2816035507, .4580167776, .6178762444, .7554044083, .8656312023, .9445750230, .9894009349};
        static const double W[8] = {.1894506104, .1826034150, .1691565193, .1495959888, .1246289712, .0951585116, .0622535239, .0271524594};
        const double XM = 0.5 * (B + A);
        const double XR = 0.5 * (B - A);
        double SS = 0.;
        for (int i = 0; i < 8; ++i) {
                double DX = XR * X[i];
                SS += W[i] * (integrand(XM + DX) + integrand(XM - DX));
        }
        return XR * SS;
}

```
In `Common.h`:
```c++
#include <functional>
(...)
double gaussInt(std::function<double(double)> integrand, double A, double B);
```

Add some tests in `testCore.cpp`, for example:
```c++
TEST(common, gaussInt)
{
        EXPECT_NEAR(gaussInt(([](double x){ return x*x; }), 0, 10), 1000/3., 1e-4);
        EXPECT_NEAR(gaussInt(([](double x){ return sin(x)*sin(x); }), 0, M_PI), M_PI/2., 1e-4);
}
```

It compiles while the test is passing :-)

Reply:
The templates are IMO the correct way here in C++, so if this works we should go for it.But please add a test to check the integration for functions with different number parameters as this is a core feature here.

Reply:
Hi @TobiasWinchen, I'm curious, why the templates approach should be the preferred way? If due to performance reasons... OK, if called frequently, I could generally agree, although, it is still ugly.

Anyway, when calling, lambda should be preferred over std::bind (since lambda is a C++ core feature, while std::bind is a library, hence, compilers can usually optimize it better). @Froehliche-Kernschmelze, in Common.h, just replace std::function with the template version you already had (and return the function body from Common.cpp):

```c++
template <typename F>
double gaussInt(F&& integrand, double A, double B) {
```
It works for me just fine.

Reply:
Thanks a real lot, everything compiles just fine now :)

Reply:
@adundovi despite all code compiling fine and actually running I find each Gauss integral to return the value of zero (the bounds A and B are properly passed). Inside `gaussInt(...)` any call to `integrand(someDouble)` will return zero. May this be related to how functions are passed to this method?

Reply:
Fixed the problem by myself (as it was a trivial implementation error by myself). Sorry for the spam. Happy to be reviewed again!